dart/lib/src/gurmukhi.dart

```diff
@@ -0,0 +1,195 @@
+  }
+
+  /// Split Gurmukhi text into letters.
+  ///
+  /// ```dart
+  /// 'ਕੈਸੀ ਆਰਤੀ ਹੋਇ ॥'.splitGurmukhi();
+  /// // ['ਕੈ', 'ਸੀ', ' ', 'ਆ', 'ਰ', 'ਤੀ', ' ', 'ਹੋ', 'ਇ', ' ', '॥']
+  /// ```
+  ///
+  /// Enable [splitVirama] to split chars joined by virama.
+  ///
+  /// ```dart
+  /// 'ਅੰਮ੍ਰਿਤ'.splitGurmukhi();
+  /// // ['ਅੰ', 'ਮ੍ਰਿ', 'ਤ']
+  /// 'ਅੰਮ੍ਰਿਤ'.splitGurmukhi(splitVirama: true);
+  /// // ['ਅੰ', 'ਮ੍', 'ਰਿ', 'ਤ']
+  /// ```
+  ///
+  /// Enable [extensions] to support SantLipi modifiers.
+  ///
+  /// ```dart
+  /// 'ਸ꠴ਯਾਮ'.splitGurmukhi(extensions: true);
+  /// // ['ਸ', '꠴ਯਾ', 'ਮ']
+  /// ```
+  Iterable<String> splitGurmukhi({
+    bool splitVirama = false,
+    bool extensions = false,
+  }) sync* {
+    var br = 0; // starting index of gurmukhi letter break
+
+    for (var i = 0; i < length; i++) {
+      final curr = codeUnitAt(i);
+      final next = i + 1 < length ? codeUnitAt(i + 1) : null;
+
+      if ((next != null && _isGurmukhiDiacritic(next)) ||
+          (!splitVirama && _isGurmukhiVirama(curr)) ||
+          (extensions && _isSantLipiYayyaModifier(curr))) {
+        continue;
+      }
+
+      yield substring(br, i + 1);
+      br = i + 1;
+    }
+  }
+
+  /// Unicode normalization of Gurmukhi text.
+  ///
+  /// Fix Matras of 'ੳ', 'ਅ', and 'ੲ':
+  /// ```dart
+  /// 'ਅਾਦਿ'.normalizeGurmukhi();
+  /// // ਆਦਿ
+  /// ```
+  ///
+  /// Fix Order of Diacritics:
+  /// ```dart
+  /// 'ਕੰੁਚਰ'.normalizeGurmukhi();
+  /// // ਕੁੰਚਰ
+  /// ```
+  String normalizeGurmukhi() {
+    return _sanitizeGurmukhi(_sortGurmukhiDiacritics(this));
+  }
+}
+
+bool _isGurmukhiChar(int char) => char >= 0x0A00 && char <= 0x0A7F;
+
+bool _isGurmukhiVirama(int char) => char == 0x0A4D;
+
+bool _isGurmukhiDiacritic(int char) => _getGurmukhiDiacriticOrder(char) != 0;
+
+bool _isSantLipiYayyaModifier(int char) {
+  return (char == 0xA833)  // ꠳
+      || (char == 0xA834)  // ꠴
+      || (char == 0xA835); // ꠵
+}
+
+int _getGurmukhiDiacriticOrder(int char) {
+  switch (char) {
+    /// nukta
+    case 0x0A3C: // ਼
+      return -2;
+
+    /// virama/halant
+    case 0x0A4D: // ੍
+      return -1;
+
+    /// signs bottom
+    case 0x0A51: // ੑ
+    case 0x0A75: // ੵ
+      return 1;
+
+    /// vowels left
+    case 0x0A3F: // ਿ
+      return 2;
+
+    /// vowels top
+    case 0x0A47: // ੇ
+    case 0x0A48: // ੈ
+    case 0x0A4B: // ੋ
+    case 0x0A4C: // ੌ
+      return 3;
+
+    /// vowels bottom
+    case 0x0A41: // ੁ
+    case 0x0A42: // ੂ
+      return 4;
+
+    /// vowels right
+    case 0x0A3E: // ਾ
+    case 0x0A40: // ੀ
+      return 5;
+
+    /// nasalization
+    case 0x0A01: // ਁ
+    case 0x0A02: // ਂ
+    case 0x0A70: // ੰ
+    case 0x0A71: // ੱ
+      return 6;
+
+    /// sign after letter
+    case 0x0A03: // ਃ
+      return 7;
+
+    /// consonant after virama
+    default:
+      return 0;
+  }
+}
+
+int _gurmukhiDiacriticComparator(int a, int b) {
+  return _getGurmukhiDiacriticOrder(a) - _getGurmukhiDiacriticOrder(b);
+}
```

Comment:
Considering the other libs use an array elements position to do the same thing, is there a reason you opted for a verbose switch here?

Reply:
I am not sure I understand you correctly, but python library uses sorting function to sort the diacritics:
https://github.com/ShabadOS/gurmukhi-utils/blob/e7d7527bcfb2a98b160a7957079a7c67694d7221/python/gurmukhiutils/unicode.py#L345C1-L354C22

In place if `sorted(key=...)`, I have used `sort(_gurmukhiDiacriticComparator)`.

Reply:
Basically instead of using a switch (which is using up many lines of code), we can just use [indexOf](https://api.dart.dev/stable/3.0.6/dart-core/List/indexOf.html) in Dart. So you could have an array called gurmukhiDiacriticOrder for example and check indexOf the two characters to get the sorting order.

```
final gurmukhiDiacriticOrder = <String>['਼', 'ੑ', 'ੵ', '੍', 'ਹ', 'ਰ', 'ਵ', 'ਟ', 'ਤ', 'ਨ', 'ਚ', 'ਿ', 'ੇ', 'ੈ', 'ੋ', 'ੌ', 'ੁ', 'ੂ', 'ਾ', 'ੀ', 'ਁ', 'ੱ', 'ਂ', 'ੰ', 'ਃ', ];
```

Effectively you're doing the same thing with the switch (iterating case by case until you find a match). This would do the same thing but with a single line of code.

The only thing you have to check is if it messes up the `_isGurmukhiDiacritic` function used in `splitGurmukhi`. As far as I can reason, it should not be a problem.

This let's you also control which characters can be used after a virama.

Reply:
I want to avoid String.indexOf() method, it will be relatively more expensive with worst case complexity O(n), what if we the user wants to convert a very long String?

I can use HashMap instead, it will reduce the number of lines, look simpler, and with O(1) lookup.

Reply:
Awesome, the performance improvement is worth the extra lines imo.

Last question: How are you going to handle consonants that are not supposed to be subscript letters? For example any letter that's not ਹ, ਰ, ਵ, ਟ, ਤ, ਨ, or ਚ?

Reply:
Is there any actual rule for that restriction? (other than ASCII fonts not having them)

AFAIK about the grammar, I think any consonant can have virama/halant added to them.

Some actual examples can even be found in Sarbloh Granth!

Reply:
Just to clarify with an example:

<img width="1377" alt="image" src="https://github.com/ShabadOS/gurmukhi-utils/assets/14130567/5f0c58a6-e8a2-4b3c-b52f-c516b8b31e72">

These words in green are not compliant per the Unicode Consortium specification. However we have support for them in Sant Lipi. 

We could add more support in Sant Lipi for the ones you find in Sarabloh Granth.

So it's not really about "rules" but about specifications. Because we cannot display the ones you're probably referencing in Sarabloh Granth (or the green ones above) with traditional fonts. Those are the ones that people have installed on their phones. So for example if someone were to try and copy paste this, it wouldn't work on iOS (I think the default font on Android does allow any consonant). We have rules to move yayya variations to a plain yayya for Unicode Consortium spec, maybe we should pick some rules for these extra consonants that they don't necessarily support.

This is more about font support rather than data, data wise we can come up with anything. And with a good enough proposal we could probably add support for these extra subscript characters into the Unicode Consortium spec, which would be awesome!

Reply:
If you have android, try this example `ਉਤ੍ਕਟ`. It is shown as a subscript in Sant Lipi here: https://shabados.github.io/SantLipi/qa/test/spg.html

This is not compliant with UC per pages 499 and 500 here https://www.unicode.org/versions/Unicode15.0.0/ch12.pdf. So I guess these are all the character combos that must be available in a unicode font (testing on macOS the ones from this PDF all worked).

However there is again the example at the top of this comment of the pairin-kakka. That did not work without Sant Lipi (I presume it would work on Android because it is included in Noto Gurmukhi).

A separate idea I had just now, is should we be normalizing ਡ + ਼  = ਡ਼ ? On the major platforms this renders as a ੜ , which I don't really know why that became the fashion. Do we want to normalize this in gurmukhi utils? Wonder what your thoughts are

Reply:
Most Unicode Gurmukhi fonts even don't support all the subscript letters you have mentioned.
Even Noto Gurmukhi fonts (default font on Android, and some Linux distros) only support ਹ, ਰ, ਵ.

In case the font doesn't support, it just shows the virama below the consonant. So, I don't think that's any issue.

See example of [Google Fonts Gurmukhi Viramas](https://fonts.google.com/?subset=gurmukhi&preview.text=%E0%A8%95%E0%A9%8D%E0%A8%B9%20%E0%A8%95%E0%A9%8D%E0%A8%B0%20%E0%A8%95%E0%A9%8D%E0%A8%B5%20%E0%A8%95%E0%A9%8D%E0%A8%9F%20%E0%A8%95%E0%A9%8D%E0%A8%A4%20%E0%A8%95%E0%A9%8D%E0%A8%A8%20%E0%A8%95%E0%A9%8D%E0%A8%9A&preview.text_type=custom).

Question: I actually don't understand what do you want to be done in case the subscript letter is not supported? You want to remove the non-supported subscript consonant, or do you want to remove the virama in that case, which will result in a different string.

Reply:
Perhaps the best solution is to allow any character as a subscript/half-form in GU, but to also add a linter/checker of sorts that points out potential issues with the string? So it could output something along the lines of `unusual character combo at position X` if it's something like the pairin-kakka.

Reply:
I was not aware about ਡ + ਼ = ਡ਼.
The solution coming to mind is that we can add a flag (0=punjabi, 1=gurmukhi) in normalization function, based on which it only normalized the nukta forms if the flag is set.

I was also thinking of adding a linter kind function, which extracts the invalid/bad letters from the strings.

Reply:
Okay, this is what I think we should do (since we have a few things going on now), please let me know your thoughts:

- Allow any VIRAMA + CONSONANT sequence (this includes things like ੍ + [ੲ, ਅ, ੳ] which is quite odd!)
- Include a linter/checker that warns odd/unusual sequences (like the virama + vowel above) and errors for malformed sequences (e.g. Halant/Udaat + Virama on the same character). This can be done in a separate PR. For the linter, I suggest the following:
  -  normal (no warning/error): ਹ, ਰ, ਵ, ਯ
  - warn (unsupported by Android, we can term them "archaic/historical"): ਟ, ਤ, ਨ, ਚ, ਠ, ਗ, ਦ, ਚ, ਮ, ਥ, ਖ
  - error: any other character
- Normalize Gurmukhi function can take a flag to normalize ਡ + ਼ = ਡ਼ (would like a specific name for the flag of sorts if possible). This also can be done in a new PR if desired.

Just providing some insight here: The reason I would like to know about any potential incorrect sequences is for example when we generate an API that could be consumed by a hukamnama bot or such (used in places like Discord, WhatsApp, Telegram, etc.). We already know what to do with yayya, I'm really not sure what we do about the unsupported pairin-akhar, but at the least we might be able to provide a note to end users that the text can't be accurately displayed and to use the website/app to see it properly (using Sant Lipi).

Reply:
I have some points:

1. How do you want handle invalid input from user? I have seen in some tests ascii+unicode mixed string is provided, which is in invalid input. What if the user inputs an Ascii letter not having any mapping? What would you like to with a string of random gurmukhi letters?
2. If you check the behaviour of the text render engine, for following two letters: ਕ੍ਰ ਕ੍ੳ, if you try to move cursor left/right on them, it behaves as they are a single letter.
3. For linter/checker: It definitely needs a different PR and proper discussion. For virama part, we can give the user an optional parameter so that list of valid subscript letters can be set.
4. Normalization flag can be named such as - nukta, nuktaNormalization, etc. I have checked on Google fonts, I believe ਡ + ਼ = ਡ਼ is more of font specific, rather than platform specific.

Reply:
1. Running toUnicode should convert any ascii that is meant to be converted to gurmukhi, and then normalize all the gurmukhi as best as it can. It's up to the user to provide the correct string. Similarly, going to ascii should only convert unicode characters and ignore the existing ascii, and then normalize all ascii sequences as best as it can (very important for the user to know how the functions work). In either function, any characters that are not meant to be changed should be left untouched.
2. That is interesting! I did not know that
3. I like the idea of adding some linter configuration to control which errors/warnings to show/hide
4. Yes, this is a font choice. It happens on android keyboards as well as iOS keyboards (because it's built into Noto Gurmukhi as well as Mukta Mahee). The flag `nuktaNormalization` makes me think it will fix normal nukta letters like ਸ + ਼ = ਸ਼, it doesn't make me think it's going to take a non-normal nukta character and change it into a ੜ. I think we need a better phrase/term.

Just sharing an image showing the font features for Noto Gurmukhi:

<img width="1055" alt="image" src="https://github.com/ShabadOS/gurmukhi-utils/assets/14130567/c929bef6-bbc0-4bbc-9a21-a6c8ad1e5826">

Reply:
The virama part, which is in sort diacritics, will only happen if user inputs a string with a virama between invalid pair of characters (I wouldn't call it a mess up from our side). As for current case, for `asciiToGurmukhi()`, this part is not a problem (ASCII fonts don't have virama mapping).

What would you actually normalize `ਡ + ਼` into? I checked `ਡ਼` isn't any Unicode character. We can try to normalize `ਡ + ਼` => `ਡ + ZWNJ + ਼`, but that again works only for some fonts (Works on Noto Sans but not on Noto Serif).

I think normalization and linter should be discussed in a separate issue, they need more research and input from others.

Reply:
It might be better to check ਡ + ਼ in the linter only. The Sant Lipi font only shows it as ਡ਼ because of the following rule:

<img width="444" alt="image" src="https://github.com/ShabadOS/gurmukhi-utils/assets/14130567/46315c1a-a92c-4d2f-bf90-dce13b1b6bed">

As you can see, we can even have characters like ਫ਼਼ (faffa with nukta, or I guess phappha with two nuktas). So it's kind of up to us how we want to render the code (just following other font conventions for now)

<img width="833" alt="image" src="https://github.com/ShabadOS/gurmukhi-utils/assets/14130567/50805b7a-b524-41ba-8e32-fdb90237d014">